### PR TITLE
UIManager: Don't block gestures for new widgets when input is disabled

### LINKED ISSUE #11117: Dialogs cause the app to be unusable when touch input is disabled
* KOReader version: v.2023.10
* Device: Onyx Boox Page (Android)

#### Issue

I use a hot corner to toggle the touch input on my device. I often read books using page turn buttons with touch input disabled to prevent accidental interactions. When touch is disabled, the "You've reached the end of the document..." dialog cannot be interacted with. Hot corners are disabled when dialogs are visible, so I can't enable touch input to interact with the dialog. Page turn buttons have no effect on the dialog. The only workaround I've found is to use system level gestures (or buttons on some devices) to exit the app. Re-opening KOReader automatically enables touch input.

This issue is applicable to all dialogs and is not exclusive to the "You've reached the end of the document..." dialog. A previous version of KOReader had a bug that would cause the "Enable WiFi" dialog to appear seemingly randomly while reading books. I had the same problem with this dialog and could only get around it by using my device's home gesture and reopening KOReader.

I can think of a 3 different solutions to this problem:
1. Touch input is enabled when dialogs are shown
2. Dialogs do not disable hot corners
3. Page turn buttons dismiss dialogs

#### Steps to reproduce

1. Setup a hot corner to toggle touch input
2. Open a book
3. Disable touch input using the hot corner
4. Use page turn buttons to complete the book
5. A dialog will popup when the book is completed

### PR DESCRIPTION
**Pull request by NiLuJe**
Fix #11117
Supersedes #11119

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11122)

### CODE REVIEW

**Comment by NiLuJe** [comment]
Draft because I need more sleep.

Also, a couple obvious footguns:

* This allows you do swap to a view where you might not have a gesture set to toggle input back on (e.g., Reader to FM via the End of Book popup).
* This doesn't account for the widget itself changing the state of the input toggle (which I guess is something that could happen?)

There's nothing we can really do about the first, but I might be able to do something about the latter, possibly with a counter instead of a bool. I need a working brain for that.

**Comment by NiLuJe** [comment]
I would have been happy with something simple like what [dfea92e](https://github.com/koreader/koreader/pull/11122/commits/dfea92e08bb1edca1b44e3e040a1cc3c43bc338e) tried to do, but, alas, we need 'em to go through at deeper levels that simply the top-level widget container.

**Comment by NiLuJe** [comment]
> * This doesn't account for the widget itself changing the state of the input toggle (which I guess is something that could happen in QuickMenu's case?)

Turns out that QM runs its callback *after* close, so it's not an issue for QM, and I don't think there's anything else that could do something like that, so I scrapped the crappier/more complex implementation that attempted to deal with it (d5f183a).

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11122)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @NiLuJe)

**Comment by NiLuJe** [comment]
Commit history archived in https://github.com/NiLuJe/koreader/tree/uimanager_and_inputcontainer_gestures (in case the more complex implementation is ever needed).

**Comment by hius07** [comment]
If a widget appears while the notification "Disabled touch input" is shown, the touch input in this widget is not enabled.
I think it's not an issue, hard to reproduce in everyday usage.

Maybe translate the notifications?

File: frontend/ui/widget/container/inputcontainer.lua [hunk 1]
```diff
@@ -304,29 +305,50 @@ end
...
+-- And the matching Event handler
+function InputContainer:onIgnoreTouchInput(toggle)
+    local Notification = require("ui/widget/notification")
+    if toggle == true then
+        if self:setIgnoreTouchInput(true) then
             Notification:notify("Disabled touch input")
         end
+    elseif toggle == false then
+        if self:setIgnoreTouchInput(false) then
+            Notification:notify("Restored touch input")
```
Comment (Frenzie) [inline frontend/ui/widget/container/inputcontainer.lua:347]: *ahem* ;-)
```suggestion
            Notification:notify(_("Restored touch input"))
```

**Comment by NiLuJe** [comment]
> If a widget appears while the notification "Disabled touch input" is shown, the touch input in this widget is not enabled.
> I think it's not an issue, hard to reproduce in everyday usage.

That's... interesting; how would you reproduce that sort of thing?

(Logically, it should go: show toast => input enabled => show whatever (input *still* enabled) => close toast => input disabled => close whatever; i.e., input should only be disabled *after* the toast disappears, which granted, if whatever isn't under a timeout, leaves it hanging ;p).

File: frontend/ui/widget/container/inputcontainer.lua (hunk 1 above)
Comment (NiLuJe) [reply frontend/ui/widget/container/inputcontainer.lua:347]: Oh, this isn't new; I thought I'd managed to magically remove gettext calls somehow :D.

**Comment by hius07** [comment]
I have a profile with "Disable touch input" action, assigned to, say, bottom right tap, that is set as "Always active".
Tap the bottom right corner to call the profile, press "Disable touch input", the profile is closed, and then immediately (until the notification is visible) tap the bottom right corner again. The profile shows up but the touch input is disabled.

**Comment by NiLuJe** [comment]
The first part makes sense and tracks with the description from my previous message. We'll live with it for now until we can find more sensible ways to break it, I think ;).

The second part is... interesting ;).
First: That onResume handler is stupid. There are a boatloads of InputContainer instances alive at any one time, so it gets called a ridonkulous amount of time. I'll move it where it belongs (DeviceListener, via the UIManager wrapper). That also means there won't ever be a notification for it, but at least that'll be by design rather than chance ;p.
Second: The ScreenSaver is (a bunch) of widgets, so it re-enables gestures during sleep. That probably screwed with the notification, as the notification only shows when there was an *effective* and non-silent state change.

**Comment by NiLuJe** [comment]
There moved to a single onResume handler, and mutilated ScreenSaver widgets to prevent UIManager from doing something stupid ;p.

**Comment by NiLuJe** [comment]
Made the ScreenSaver stuff slightly less awful.

**Review by NiLuJe — COMMENTED** [review]
> Reviewed 2 of 3 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/11122)* status: all files reviewed, 1 unresolved discussion (waiting on @Frenzie)

**Comment by hius07** [comment]
> After that I enable touch input by sleep/wakeup (by the cover), and if I call this profile and press "Disable touch input" - the notification does not appear.

I must add that if I press "Enable touch input" (I have this action in that QM as well), the "enabled" notification appears. After that the "disabled" notification appears again in normal way.

**Comment by NiLuJe** [comment]
> I must add that if I press "Enable touch input" (I have this action in that QM as well), the "enabled" notification appears. After that the "disabled" notification appears again in normal way.

Yup, that's good.

Note that there won't be a notification on resume with the updated code, but at least it's by design rather than by chance this time ;).

**Pull request merged**